### ISSUE TITLE
long running fetchmail blocks cron

### ISSUE THREAD

**Issue by yelizariev** [odoo/odoo#7464]
Sometimes fetchmail run really long (10-20 hours):

```
2015-07-04 18:31:40,019 20702 DEBUG eg-prices.com openerp.addons.base.ir.ir_cron: 26491.446s (fetchmail.server, _fetch_mails)

2015-07-05 21:04:40,312 29476 DEBUG eg-prices.com openerp.addons.base.ir.ir_cron: 31459.958s (fetchmail.server, _fetch_mails)

2015-07-06 06:15:15,060 29476 DEBUG www.it-projects.info openerp.addons.base.ir.ir_cron: 63809.573s (fetchmail.server, _fetch_mails)
```

It completely blocks cron, to say nothing of fetching mails.

I think some timeout shoud be set, e.g. [socket.setdefaulttimeout](https://docs.python.org/2/library/socket.html#socket.setdefaulttimeout)

**Comment by yelizariev** [comment]
I use odoo 8.0

    # git log -n 1
    commit 0f01df42eaf662987abe5f069ca8df2e2caf8e37
    Merge: 114e95b f15cbd6
    Author: Olivier Dony <odo@openerp.com>
    Date:   Thu Jun 12 18:59:15 2014 +0200

        [MERGE] Forward-port saas-5 up to f15cbd6

**Comment by Yenthe666** [comment]
I've experienced this problem on some Odoo's too. In fact this seems to be the case for multiple ir.cron tasks. Somehow every once in a while they seem to run stuck.
I honestly think that a time out would be a good fix. If it takes unusually long it should time-out and should then try again.

**Comment by rafaelbn** [comment]
cc @pedrobaeza @antespi @Yajo

**Comment by rafaelbn** [comment]
@yelizariev it's much more effective if you make a PR to Odoo fixing it.
Also you can make a PR to https://github.com/OCA/OCB fixing the problem and have it solve it
:+1:

**Comment by pedrobaeza** [comment]
This is already done: https://github.com/odoo/odoo/blob/8.0/addons/fetchmail/fetchmail.py#L159, so it can be closed.

**Comment by elicoidal** [comment]
@pedrobaeza This is actually not enough.

We have created a patch this week that actually works and we will submit it next week to Odoo.
This current fix actually fixed the TIMEOUT after the connection is established while you need to fix it for establishing the connection (you need to change the method IMAP4_SSL and add a timeout parameter).

Our experience in China: with current fix we had continuous times out. Since we have this fix (3 or 4 days), we have no issue anymore.
We have to package a fix and submit to Odoo or if refused prepare a module for the OCA

**Comment by pedrobaeza** [comment]
OK, let me know

**Comment by elicoidal** [comment]
cc @seb-elico

**Comment by daramousk** [comment]
@elicoidal Has a final fix been provided on this ?

**Comment by elicoidal** [comment]
@daramousk We made a PR which was not accepted: https://github.com/odoo/odoo/pull/9884

**Comment by pedrobaeza** [comment]
@elicoidal maybe you can prepare a similar patch for v10/v11 or master. Any way, I see you are covering only IMAP, but not POP3.

**Comment by elicoidal** [comment]
@seb-elico could you have a look to propose a patch.
@pedrobaeza patch on v8 was not accepted: do you have any idea why? If we redo similar patch we might end up in the same situation

**Comment by Yenthe666** [comment]
@elicoidal the PR was denied & automatically closed because at that point Odoo 8 was no longer official supported.
In short: nobody really looked at your PR (automated message) and somebody should have a look when you make it against 9/10/11. Sorry for that closure 😞

**Comment by elicoidal** [comment]
@Yenthe666 Well noted. We will have a look for v10 for the moment which is our running version

**Comment by Yenthe666** [comment]
Perfect. Let me know when you've made a PR. 😉

**Comment by Yenthe666** [comment]
@seb-odoo don't forget the guidelines. If its a big you can create a PR for the lowest supported version where you have it in. If it is a new feature or addition in Odoo it should be only against master.

**Comment by yelizariev** [comment]
> If its a b**a**g

**Comment by seb-elico** [comment]
> If its a b**u**g

@Yenthe666 Thanks for the reminder. I'm not the one to judge if not having a timeout is a bug though. IMO, adding a timeout is a new feature. Even if, in this case, it's to avoid having the cron run indefinitely, which could be considered as a bug. It's a question of interpretation. I would appreciate to see this patch back-ported obviously!

**Comment by rafaelbn** [comment]
I will tell my grandchildren about this :smile: 

There was a BUG and a contributor made the PR in 2015 https://github.com/odoo/odoo/pull/9884 but one (1) year after, author close it because the version was not supported!!! OMG!

Maybe if @yelizariev would have made the PR in July 2015 Odoo would have merged and not closed? :laughing:

**Comment by daramousk** [comment]
I see that a fix has already been merged on 54a477d. Perhaps this issue should be closed?

**Comment by Yenthe666** [comment]
Yep, I believe that should fix it indeed @daramousk! Thanks for letting me know 😉

**Comment by pjrobertson** [comment]
I know this issue is old, but I just want to put in my 2 cents in case it's useful for anybody Googling this.

My situation is that I have an Odoo server inside China, and an IMAP mail server outside of China. So again, we're dealing with China's Great Firewall issues...

I was having issues with the HR (Recruitment) addon picking up new emails from a catch-all email address. The emails were arriving in the catch-all mailbox, and were marked as read. And the Technical Settings > Incoming Mail Servers screen was showing that the cron job was running correctly (Last Fetch Date was updating correctly).

After a bit more investigation of the logs, I realised that the Odoo fetchmail service was logging into the mailbox, starting to download messages, and then timing out. This meant messages were getting marked as 'read' but not being fully picked up by Odoo.

My `limit_time_real` in odoo.conf was originally 120s. I upped this to 300s and now 3600s (1 hour!!). That seems to have done the trick, and seems fairly similar to the issues and fix @elicoidal was facing.

I'm not sure how IMAP works, but if there was a way to only mark a message as read *after* it's fully been picked up and parsed by Odoo, that could be one step in avoiding this issue. The flow would be something like:

Connect to IMAP server → Check mail.. 1 new message → Download message → Parse message → Mark message as read on IMAP server → Close.
